File: web/src/gf_apps/gf_crawl_lib/ts/gf_crawl_stats/gf_crawl_stats__fetches.ts
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

declare var c3;
//---------------------------------------------------
export function view__fetches_per_day(p_fetches_by_days_map, p_parent, p_log_fun) {
    p_log_fun('FUN_ENTER','gf_crawl_stats__fetches.view__fetches_per_day()');

    const plot = $(`
        <div id='plots'>
            <div id='new_fetches_per_day__plot'>
                <svg width='2000' height='2000'></svg>
            </div>
        </div>`);

    $(p_parent).append(plot);

    const counts_by_day__sorted_lst        = p_fetches_by_days_map['counts_by_day__sorted_lst'];
    const domain_counts_by_day__sorted_lst = p_fetches_by_days_map['domain_counts_by_day__sorted_lst'];

    /*//------------------
    const daily_count_lst            = [];
    const daily_count_per_domain_map = {}; //:Map<:List<:Int>>

    for (var day_stat_map of p_stats_lst) {

    	const count_int                   = day_stat_map['total_count_int'];
        const total_count__per_domain_map = day_stat_map['total_count__per_domain_map'];


    	daily_count_lst.push(count_int);

        //accumulate up domain counts per day into a specific domains counts lists,
        for (var domain_str in total_count__per_domain_map) {

            const domain_count_int = total_count__per_domain_map[domain_str];

            if (domain_str in daily_count_per_domain_map) {
                daily_count_per_domain_map[domain_str].push(domain_count_int);
            } else {
                daily_count_per_domain_map[domain_str] = [domain_count_int];
            }
        }
    }*/
    //------------------
    //C3_COLUMNS

    counts_by_day__sorted_lst.unshift('per-day fetches count');

    const c3_columns_lst = [
        counts_by_day__sorted_lst
    ]

    for (var domain_counts_map of domain_counts_by_day__sorted_lst) {

        const domain_str      = domain_counts_map['domain_str'];
        const days_counts_lst = domain_counts_map['days_counts_lst'];
        //const domain_counts_lst = daily_count_per_domain_map[domain_str];

        //create a title of the columns for C3
        days_counts_lst.unshift(domain_str);

        c3_columns_lst.push(days_counts_lst);
    }
    //------------------

    const top_c3_columns_lst = c3_columns_lst.slice(0,20);

    console.log(top_c3_columns_lst)

    const chart = c3.generate({
        bindto: '#new_fetches_per_day__plot',
        data: {
            columns: top_c3_columns_lst
            /*columns: [
            daily_total_count_lst,
            //['data1', 30, 200, 100, 400, 150, 250],
            //['data2', 50, 20, 10, 40, 15, 25]
            ]*/
        }
    });
}

File: go/gf_apps/gf_crawl_lib/gf_crawl_stats/gf_crawl_stats__fetches.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_crawl_stats

import (
	"context"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/bson"
	"github.com/gloflow/gloflow/go/gf_core"
)

//-------------------------------------------------

type Gf_stat__crawled_url_fetches struct {
	Url_str         string    `bson:"_id"             json:"url_str"`
	Count_int       int       `bson:"count_int"       json:"count_int"`
	Start_times_lst []float64 `bson:"start_times_lst" json:"start_times_lst"`
}

//-------------------------------------------------

func stats__crawler_fetches_by_days(pRuntimeSys *gf_core.RuntimeSys) (map[string]interface{}, *gf_core.GFerror) {
	pRuntimeSys.LogFun("FUN_ENTER", "gf_crawl_stats__fetches.stats__crawler_fetches_by_days()")

	stats__fetches_by_days, gf_err := stats__objs_by_days(map[string]interface{}{}, "crawler_url_fetch", pRuntimeSys)
	if gf_err != nil {
		return nil, gf_err
	}

	data_map := map[string]interface{}{
		"fetches_by_days_map": stats__fetches_by_days,
	}
	return data_map, nil
}

//-------------------------------------------------

func stats__crawler_fetches_by_url(pRuntimeSys *gf_core.RuntimeSys) (map[string]interface{}, *gf_core.GFerror) {
	pRuntimeSys.LogFun("FUN_ENTER", "gf_crawl_stats__fetches.stats__crawler_fetches_by_url()")


	ctx := context.Background()
	pipeline := mongo.Pipeline{
		{
			{"$match", bson.M{"t": "crawler_url_fetch"}},
		},
		{
			{"$project", bson.M{
				"id_str":       true,
				"start_time_f": true,
				"end_time_f":   true,
				"domain_str":   true,
				"url_str":      true, // actual link from the HTML <a> page ('href' parameter)
				// "errors_num_i":bson.M{"$size":"$errors_lst",},
			}},
		},
		{
			{"$group", bson.M{
				"_id":             "$url_str",
				"count_int":       bson.M{"$sum":  1},
				"start_times_lst": bson.M{"$push": "$start_time_f"},
			}},
		},
		{
			{"$sort", bson.M{"count_int": -1}},
		},
	}


	/*pipe := pRuntimeSys.Mongo_db.Collection("gf_crawl").Pipe([]bson.M{
		bson.M{"$match": bson.M{
				"t":     "crawler_url_fetch",
			},
		},

		bson.M{"$project": bson.M{
				"id_str":       true,
				"start_time_f": true,
				"end_time_f":   true,
				"domain_str":   true,
				"url_str":      true, // actual link from the HTML <a> page ('href' parameter)
				// "errors_num_i":bson.M{"$size":"$errors_lst",},
			},
		},

		bson.M{"$group": bson.M{
				"_id":             "$url_str",
				"count_int":       bson.M{"$sum":  1},
				"start_times_lst": bson.M{"$push": "$start_time_f"},
			},
		},

		bson.M{"$sort": bson.M{
				"count_int": -1,
			},
		},
	})*/

	cursor, err := pRuntimeSys.Mongo_db.Collection("gf_crawl").Aggregate(ctx, pipeline)
	if err != nil {

		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to group all crawler_url_fetch's",
			"mongodb_aggregation_error",
			map[string]interface{}{},
			err, "gf_crawl_stats", pRuntimeSys)
		return nil, gf_err
	}
	defer cursor.Close(ctx)

	/*results_lst := []Gf_stat__crawled_url_fetches{}
	err         := pipe.All(&results_lst)

	if err != nil {
		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to group all crawler_url_fetch's",
			"mongodb_aggregation_error",
			nil, err, "gf_crawl_stats", pRuntimeSys)
		return nil, gf_err
	}*/
	
	results_lst := []Gf_stat__crawled_url_fetches{}
	for cursor.Next(ctx) {

		var r Gf_stat__crawled_url_fetches
		err := cursor.Decode(&r)
		if err != nil {
			gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to group all crawler_url_fetch's",
				"mongodb_cursor_decode",
				map[string]interface{}{},
				err, "gf_crawl_stats", pRuntimeSys)
			return nil, gf_err
		}
		results_lst = append(results_lst, r)
	}

	data_map := map[string]interface{}{
		"crawled_url_fetches_lst": results_lst,
	}

	return data_map, nil
}